### DISCUSSION TITLE
Trouble with empty array

### DISCUSSION THREAD

**Discussion by ymc9** [eclipse-langium/langium#975] [Q&A] [answered]
Hi there,

I'm running into a parsing problem with my array syntax. Here's a simplified repro:

```langium
grammar HelloWorld

entry Model:
    (persons+=Person | greetings+=Greeting)*;

Person:
    'person' name=ID (array?='[]')?;

Array: '[' (items+=ID)* ']';

Greeting:
    'Hello' person=[Person:ID] items=Array '!';

hidden terminal WS: /\s+/;
terminal ID: /[_a-zA-Z][\w_]*/;
```

Input:
```
person John
Hello John []!
```

Error (reproduced on langium playground):
```
Expecting keyword '[' but found `[]`.HelloWorld(parsing-error)
Expecting keyword '!' but found `[]`.HelloWorld(parsing-error)
```

The empty array on the second line wouldn't parse, and it seems the conflict is introduced by the `(array?='[]')` in `Person` rule. Does such a rule generate a token implicitly?

**Comment by msujew** [comment, marked as answer]
Hey @ymc9,

> Does such a rule generate a token implicitly?

Yes, every unique usage of keywords (anything wrapped inside of a string) generates a token type for the lexer. This is expected behavior for basically all parser generators (such as ANTLR). The obvious fix is to modify the `Person` rule slightly:

```
Person:
    'person' name=ID (array?='[' ']')?;
```

The issue is that since `[]` is a longer token type pattern, it has precedence over shorter ones, such as `[`. So during the lexing phase, encountering the string `[]` will always yield the `[]` token instead of separate `[` and `]` tokens.

**Reply by ymc9** [reply]
Got it. Thanks for the confirmation and solution!
